spec/rubocop/cop/lint/number_conversion_spec.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+RSpec.describe RuboCop::Cop::Lint::NumberConversion do
+  subject(:cop) { described_class.new(config) }
+
+  let(:config) { RuboCop::Config.new }
+
+  it 'registers an offense when using `#to_i`' do
+    expect_offense(<<-RUBY.strip_indent)
+      '10'.to_i
+      ^^^^^^^^^ Replace unsafe number conversion with number class parsing, instead of using "10".to_i, use stricter Integer("10").
+    RUBY
+  end
+
+  it 'registers an offense when using `#to_i` for integer' do
+    expect_offense(<<-RUBY.strip_indent)
+      10.to_i
+      ^^^^^^^ Replace unsafe number conversion with number class parsing, instead of using 10.to_i, use stricter Integer(10).
+    RUBY
+  end
```

Comment:
In addition, It would be good to add a practical test case calling `to_i` method to variables.

Examples: `foo.to_i`, `params[:foo].to_i`

These test cases need to return the expected message.

Reply:
I've added an test scenario, which test cop against a variable.

Reply:
I don't know how do I get callee method/variable. I used to do Like extracting `params[:foo]` from `params[:foo].to_i`. Is there any easier way I can get this from a node other than regex or something ?

Reply:
For example, using capture of `NodePattern` is a simple solution.
https://github.com/bbatsov/rubocop/blob/v0.52.1/lib/rubocop/node_pattern.rb#L35

Here is an example using capture with `def_node_matter` (a little refactoring is included) 📝 


```diff
diff --git a/lib/rubocop/cop/lint/number_conversion.rb b/lib/rubocop/cop/lint/number_conversion.rb
index f7984b3af..23c215379 100644
--- a/lib/rubocop/cop/lint/number_conversion.rb
+++ b/lib/rubocop/cop/lint/number_conversion.rb
@@ -21,65 +21,38 @@ module RuboCop
       #   Float('10.2')
       #   Complex('10')
       class NumberConversion < Cop
-        NUMBER_CONVERSION_METHOD_CLASS_MAPPING = {
+        CONVERSION_METHOD_MAPPINGS = {
           to_i: Integer.name,
           to_f: Float.name,
           to_c: Complex.name
         }.freeze
+
         MSG = 'Replace unsafe number conversion with number '\
               'class parsing, instead of using '\
-              '%{warning_method}, use stricter '\
-              '%{number_class}(%{input_value}).'.freeze
-
-        def_node_matcher :to_i?, <<-PATTERN
-          (send _recv :to_i)
-        PATTERN
+              '%{numeric_object}.%{to_method}, use stricter '\
+              '%{corrected_method}(%{numeric_object}).'.freeze

-        def_node_matcher :to_f?, <<-PATTERN
-          (send _recv :to_f)
-        PATTERN
-
-        def_node_matcher :to_c?, <<-PATTERN
-          (send _recv :to_c)
+        def_node_matcher :to_method, <<-PATTERN
+          (send $_ ${:to_i :to_f :to_c})
         PATTERN

         def on_send(node)
-          return unless node_violate_rules?(node)
-          @node = node
-
-          message = format(
-            MSG,
-            input_value: input_value,
-            warning_method: warning_method,
-            violating_method: violating_method,
-            number_class: number_class
-          )
-          add_offense(node, message: message)
+          to_method(node) do |receiver, to_method|
+            message = format(
+              MSG,
+              numeric_object: receiver.source,
+              to_method: to_method,
+              corrected_method: correct_method(node.method_name)
+            )
+
+            add_offense(node, message: message)
+          end
         end

         private

-        def node_violate_rules?(node)
-          to_i?(node) ||
-            to_f?(node) ||
-            to_c?(node)
-        end
-
-        def input_value
-          input_value = @node.children.first.children.first
-          input_value.is_a?(Symbol) ? input_value.to_s : input_value.inspect
-        end
-
-        def warning_method
-          @node.source
-        end
-
-        def violating_method
-          @node.method_name
-        end
-
-        def number_class
-          NUMBER_CONVERSION_METHOD_CLASS_MAPPING[@node.method_name]
+        def correct_method(method_name)
+          CONVERSION_METHOD_MAPPINGS[method_name]
         end
       end
     end
diff --git a/spec/rubocop/cop/lint/number_conversion_spec.rb b/spec/rubocop/cop/lint/number_conversion_spec.rb
index 10a12692f..95a5bec00 100644
--- a/spec/rubocop/cop/lint/number_conversion_spec.rb
+++ b/spec/rubocop/cop/lint/number_conversion_spec.rb
@@ -43,9 +43,8 @@ RSpec.describe RuboCop::Cop::Lint::NumberConversion do

   it 'registers an offense when `#to_i` called on a variable' do
     expect_offense(<<-RUBY.strip_indent)
-      params = { id: 10 }
       params[:id].to_i
-      ^^^^^^^^^^^^^^^^^ Replace unsafe number conversion with number class parsing, instead of using params[:id].to_i, use stricter Integer(params[:id]).
+      ^^^^^^^^^^^^^^^^ Replace unsafe number conversion with number class parsing, instead of using params[:id].to_i, use stricter Integer(params[:id]).
     RUBY
   end
```

Reply:
@koic Spot on, I think I should read `node_pattern`, I was referring `readthedocs` as a guide, it doesn't have much. Thank you 👏.